metricbeat/scripts/mage/docs_collector.go

```diff
@@ -0,0 +1,413 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package mage
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path"
+	"path/filepath"
+	"sort"
+	"strings"
+
+	"github.com/pkg/errors"
+	"gopkg.in/yaml.v2"
+
+	"text/template"
+
+	"github.com/elastic/beats/dev-tools/mage"
+)
+
+// moduleData provides module-level data that will be used to populate the module list
+type moduleData struct {
+	Path       string
+	Base       string
+	Title      string `yaml:"title"`
+	Release    string `yaml:"release"`
+	Dashboards bool
+	Settings   []string `yaml:"settings"`
+	CfgFile    string
+	Asciidoc   string
+	Metricsets []metricsetData
+}
+
+type metricsetData struct {
+	Path       string
+	Title      string
+	Link       string
+	Release    string
+	DataExists bool
+}
+
+func writeTemplate(filename string, t *template.Template, args interface{}) error {
```

Comment:
Would it make sense to reuse `mage.Expand...` functions?

Reply:
I was looking at those, and they didn't seem particularly helpful, as they wanted a map and the module list needed to be sorted. I would just end up doing some hack with a top-level map and an array.

Reply:
The expand functions are also a tad weird, and don't really give me an elegant way to pass it multiple template files like you can with `ParseGlob`. This way just feels a tad 'cleaner' since I only have to call `Parse` once and just pass a template variable to all the writers?

Reply:
:+1: